apbs: Update to current 3.4.1

jwhowarth opened the pull request:
apbs: Update the apbs package to the current 3.4.1 release using Debian's approach and patches but with building the apbs python module which is unnecessary for use of apbs in pymol.

#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
macOS 14.4.1 23E224 arm64
Xcode 15.3 15E204a

###### Verification <!-- (delete not applicable items) -->
Have you

- [x ] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x ] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x ] checked your Portfile with `port lint --nitpick`?
- [x ] tried existing tests with `sudo port test`?
- [ x] tried a full install with `sudo port -vst install`?
- [x ] tested basic functionality of all binary files?
- [x ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

jwhowarth commented:
Does anyone have access to macos-11 or macos-12 or the actual build logs for the failing builds. I don't have the ability to test either of those here to sort out why the use of a python_select3 dependency is failing on the buildbot. Again, my aim was to avoid needlessly creating a set of python variants for apbs on darwin prior to 13 (when python3 was added to the command tools) since it only needs a python3 executable to run the testsuite. Actually use of apbs by pymol can be entirely done from the apbs executable without additional python support.

reneeotten commented on science/apbs/Portfile:
```diff
@@ -20,64 +18,75 @@ long_description        APBS is a software package for the numerical \
-# dependencies are not managed correctly within this code, so errors such as from -lapbs_geoflow might occur
-# how to solve: first 'make apbs_geoflow', then 'make all'
-use_parallel_build no
+use_parallel_build      yes
```
this is the default, so just remove the line instead of setting it to the default value

reneeotten commented on science/apbs/Portfile:
```diff
@@ -20,64 +18,75 @@ long_description        APBS is a software package for the numerical \
-    }
-    
-    system -W ${worksrcpath}/tests "python apbs_tester.py"
+    system -W ${worksrcpath}/tests "PATH=${destroot}${prefix}/bin:/usr/bin:/bin:${prefix}/bin python3 apbs_tester.py -c test_cases.cfg || exit 0"
```
see above, don't rely on being `python3` to be present, instead add a dependency on MacPorts' default Python port and use that

reneeotten commented on science/apbs/Portfile:
```diff
@@ -20,64 +18,75 @@ long_description        APBS is a software package for the numerical \
+    platform darwin {
+       if {${os.major} < 13} {
+          reinplace "s|env python3|env ${prefix}/bin/python3|" ${worksrcpath}/tools/manip/psize.py
+          reinplace "s|/usr/bin/python3|${prefix}/bin/python3|" ${worksrcpath}/tools/python/vgrid/mergedx.py
```
as above

jwhowarth commented:
I seem to be unable to see the detailed build logs for macos-11 or macos-12 to see exactly what is going wrong with the requested changes being added. I thought cmake would find python3.12 on its own as I noticed that when python3 on darwin13 is set to python3.11, cmake appears to go for the highest python it can find...

//Details about finding Python3
FIND_PACKAGE_MESSAGE_DETAILS_Python3:INTERNAL=[/opt/local/bin/python3.12][cfound components: Interpreter ][v3.12.3(3.6)]
_Python3_Compiler_REASON_FAILURE:INTERNAL=
_Python3_Development_REASON_FAILURE:INTERNAL=
_Python3_EXECUTABLE:INTERNAL=/opt/local/bin/python3.12

reneeotten commented on science/apbs/Portfile:
```diff
@@ -20,64 +18,81 @@ long_description        APBS is a software package for the numerical \
+# is unnecessary for apbs's use in pymol.
+platform darwin {
+   if {${os.major} < 13} {
+       depends           port:port:python3.12
```
only using "depends" here can surely not work....

jwhowarth commented on science/apbs/Portfile:
```diff
@@ -3,8 +3,6 @@
-PortGroup               mpi 1.0
-PortGroup               github 1.0
 
 cmake.out_of_source     yes
```
Note sure exactly which line you are referring to here... 

PortSystem              1.0
PortGroup               compiler_blacklist_versions 1.0
make.out_of_source     yes

reneeotten replied:
well the comment is shown for kine 7, so that's the one I am referring to 😉. Once you're using the latest "cmake" PortGrout the "out_of_source" thing is the default

jwhowarth commented:
The usage of apbs in pymol can be tested by installing this apbs package as well as the current pymol and pdb2pqr packages. Down a pub to load such as [https://files.rcsb.org/download/1TN4.pdb](https://files.rcsb.org/download/1TN4.pdb). From a terminal window, execute 'pymol' and the select the Open command from the File menu in the pymol GUI to select the 1TN4.pdb file. Once pymol displays the molecule, select the  APBS Electrostatic menu item from the Plugins menu. In the APBS Electrostatics dialog that appears, click on the Run button. When the alert appears asking if the APBS Electrostatics dialog can be dismissed, click the Yes button. You should see a blue. red and white electrostatic surface now displayed over the molecule.

reneeotten commented on science/apbs/Portfile:
```diff
@@ -20,64 +18,85 @@ long_description        APBS is a software package for the numerical \
-github.setup            Electrostatics apbs-pdb2pqr d4e78c62e6a07f92c26924318d83a159fa6af063
-version                 1.4.1
+version                 3.4.1
 homepage                http://www.poissonboltzmann.org
```
please use "https"

reneeotten commented on science/apbs/Portfile:
```diff
@@ -20,64 +18,85 @@ long_description        APBS is a software package for the numerical \
+# is unnecessary for apbs's use in pymol.
+platform darwin {
+   if {${os.major} < 13} {
+       depends_build     port:port:python312
```
The comment suggests Python is only needed for the tests, so shouldn't it be a test dependency then?

jwhowarth replied:
No, a depend_build is required as the build requires cmake to find a python3 in order to satisfy the test requirements. In the absence of a python3, the cmake call will fail during the build itself. Definitely annoy but I don't see it as a big deal.

jwhowarth commented:
I tried disabling CMakefiles.txt from looking for python by commenting out...

set(Python3_USE_STATIC_LIBS TRUE)
set(Python3_USE_STATIC_LIBS FALSE)

and the cmake process still looks for python3 due to the test harness being in python.

-- ******* TESTING ADDED *****************
-- Found Python3: /opt/local/bin/python3.12 (found suitable version "3.12.3", minimum required is "3.6") found components: Interpreter

I don't think it is worth hacking the CMakefiles in a way that may become fragile in later releases just to avoid having python (required for testing) found at build time as well.

jwhowarth commented:
The latest commit appears to have successfully suppress the unused check on python3 in the build so we can just rely on the depends_test for python312 on Darwin earlier than 13.

jwhowarth commented:
I believe I had addressed all of the issues at this point and the pull should be ready for commit. After it is committed, I'll start working on updating the pdb2pqr package to take it to the matching 3.4.1 release and convert that packing to python variants as pdb2pqr was rewritten to be entirely python code rather a mix of python and compile c code in the older version currently in macports.

jwhowarth commented:
Am I missing any requested changes that are blocking the commit? I don't see any obvious ones left.

jwhowarth commented:
Will replace with a clean pull request in hopes of more progress towards a commit.

blair commented:
Where is the new pull request?

reneeotten commented:
Closing this PR and opening a new one will not speed anything up. In fact, it's a bad idea as all the discussion/context will be lost in a new attempt. This is a volunteer project and people including me will do things as time permits; currently, my time does not permit so in whatever PR (new or old) you'll have to wait until people will take a look.
